File: netDxf/Objects/DxfStoredTableGeometry.Edit.cs
// Copyright (c) netDxf contributors. Licensed under the MIT License.
using System;
using System.Collections.Generic;
using System.Linq;
using netDxf.IO;

namespace netDxf.Objects
{
    public sealed partial class DxfStoredTableGeometry
    {
        private bool editing;
        private bool reentered;

        /// <summary>Atomically replaces the qualified stored geometry packet of this loaded object.</summary>
        /// <param name="rowCount">The nonnegative stored row count, at most 1,048,576.</param>
        /// <param name="columnCount">The nonnegative stored column count, at most 1,048,576.</param>
        /// <param name="cells">The ordered immutable cell values. Null entries are not permitted.</param>
        /// <remarks>
        /// The complete request is copied and validated after caller enumeration and disposal.
        /// References must be explicit actual registered identities in this source document,
        /// including retained owner-held metadata; no handles are assigned and foreign objects
        /// are never matched by name or handle. A request equivalent to the current values keeps
        /// the original packet. Earlier Payload, Cells and References snapshots remain unchanged.
        /// Counts are encoded from the supplied lists, without inferring row/column addresses or
        /// requiring row count times column count to equal cell count. Negative finite dimensions
        /// and uninterpreted signed flags remain allowed. The complete emitted record is limited
        /// to 1,048,576 tags. The source profile, owner graph and common metadata are retained.
        /// This operation does not evaluate layout, regenerate a TABLE, create or erase objects,
        /// or roll back independent document changes made by caller enumeration callbacks.
        /// </remarks>
        public void ReplaceGeometry(int rowCount, int columnCount, IEnumerable<DxfStoredTableGeometryCell> cells)
        {
            if (this.editing)
            {
                this.reentered = true;
                throw new InvalidOperationException("TABLEGEOMETRY replacement cannot be reentered.");
            }
            if (cells == null) throw new ArgumentNullException(nameof(cells));
            if (rowCount < 0 || rowCount > MaximumPayloadTags) throw new ArgumentOutOfRangeException(nameof(rowCount));
            if (columnCount < 0 || columnCount > MaximumPayloadTags) throw new ArgumentOutOfRangeException(nameof(columnCount));
            this.editing = true;
            this.reentered = false;
            try
            {
                long tagCount = 4;
                var replacement = new List<DxfStoredTableGeometryCell>();
                foreach (DxfStoredTableGeometryCell cell in cells)
                {
                    if (cell == null) throw new ArgumentException("A stored geometry cell cannot be null.", nameof(cells));
                    tagCount += 5L + 11L * cell.Geometry.Count;
                    if (tagCount > MaximumPayloadTags)
                        throw new ArgumentException("The requested geometry exceeds the record tag limit.", nameof(cells));
                    replacement.Add(cell);
                }
                // The enumerator has been disposed. All remaining work uses sealed values and
                // document-owned collections and cannot invoke caller enumeration again.
                if (this.reentered) throw new InvalidOperationException("TABLEGEOMETRY replacement was reentered during enumeration.");
                this.ValidateReplacementSource();
                if (this.StoredRecordTagCount(tagCount) > MaximumPayloadTags)
                    throw new ArgumentException("The requested geometry and common metadata exceed the record tag limit.", nameof(cells));

                var nextReferences = new List<DxfObject>();
                var nextHandles = new Dictionary<string, DxfObject>(StringComparer.OrdinalIgnoreCase);
                foreach (DxfStoredTableGeometryCell cell in replacement)
                {
                    DxfObject target = cell.GeometryReference;
                    if (target == null) continue;
                    if (target.Handle == null || !ReferenceEquals(this.source.StoredTableHandleTarget(target.Handle), target))
                        throw new ArgumentException("Every geometry reference must be an actual registered identity in the source document.", nameof(cells));
                    nextReferences.Add(target);
                    nextHandles[ReferenceText(cell)] = target;
                }
                if (this.SameGeometry(rowCount, columnCount, replacement)) return;

                var tags = new List<DxfTag>((int)tagCount)
                {
                    new DxfTag(100, "AcDbTableGeometry"), new DxfTag(90, rowCount),
                    new DxfTag(91, columnCount), new DxfTag(92, replacement.Count)
                };
                foreach (DxfStoredTableGeometryCell cell in replacement)
                {
                    tags.Add(new DxfTag(93, cell.GeometryDataFlags));
                    tags.Add(new DxfTag(40, cell.WidthWithGap));
                    tags.Add(new DxfTag(41, cell.HeightWithGap));
                    tags.Add(new DxfTag(330, ReferenceText(cell)));
                    tags.Add(new DxfTag(94, cell.Geometry.Count));
                    foreach (DxfStoredTableCellGeometry value in cell.Geometry)
                    {
                        AddPoint(tags, 10, value.TopLeftDistance); AddPoint(tags, 11, value.CenterDistance);
                        tags.Add(new DxfTag(43, value.ContentWidth)); tags.Add(new DxfTag(44, value.ContentHeight));
                        tags.Add(new DxfTag(45, value.Width)); tags.Add(new DxfTag(46, value.Height));
                        tags.Add(new DxfTag(95, value.StoredValue95));
                    }
                }
                var packet = tags.AsReadOnly();
                var cellSnapshot = replacement.AsReadOnly();
                // Every callback, allocation and validation has finished. Only state swaps remain.
                this.Payload = packet;
                this.Cells = cellSnapshot;
                this.references = nextReferences;
                this.handles = nextHandles;
                this.RowCount = rowCount;
                this.ColumnCount = columnCount;
            }
            finally { this.editing = false; this.reentered = false; }
        }

        private void ValidateReplacementSource()
        {
            if (!this.resolved || this.IsErased || this.Database == null
                || !ReferenceEquals(this.Database.Document, this.source)
                || !ReferenceEquals(this.source.GetObjectByHandle(this.Handle), this))
                throw new InvalidOperationException("TABLEGEOMETRY must remain registered in its source document.");
            // This read-only validation includes source ownership, root dictionary links,
            // declared wrapper slots, dependencies, metadata and every owner ancestor.
            IReadOnlyList<string> errors = this.Database.Validate();
            if (errors.Count != 0)
                throw new InvalidOperationException("Cannot replace TABLEGEOMETRY in an invalid source database: " + string.Join("; ", errors));
        }

        private long StoredRecordTagCount(long payloadCount)
        {
            long count = payloadCount + 2; // Identity and common owner, excluding group 0.
            if (this.ExtensionDictionary != null) count += 3;
            int reactors = this.PersistentReactors.Where(item => item != null).Select(item => item.Handle).Distinct(StringComparer.OrdinalIgnoreCase).Count();
            if (reactors != 0) count += reactors + 2L;
            foreach (XData data in this.XData.Values)
            {
                count++; // Application registry marker.
                foreach (XDataRecord record in data.XDataRecord)
                    count += record.Code == XDataCode.BinaryData ? Math.Max(1L, (((byte[])record.Value).LongLength + 126L) / 127L) : 1L;
            }
            return count;
        }

        private static string ReferenceText(DxfStoredTableGeometryCell cell)
        {
            string current = cell.GeometryReference == null ? "0" : cell.GeometryReference.Handle;
            // Keep source lexical handle spelling when retaining a loaded cell identity.
            if (cell.ReferenceHandle != null && Convert.ToUInt64(cell.ReferenceHandle, 16) == Convert.ToUInt64(current, 16))
                return cell.ReferenceHandle;
            return current;
        }

        private bool SameGeometry(int rows, int columns, IList<DxfStoredTableGeometryCell> cells)
        {
            if (rows != this.RowCount || columns != this.ColumnCount || cells.Count != this.Cells.Count) return false;
            for (int i = 0; i < cells.Count; i++)
            {
                DxfStoredTableGeometryCell a = cells[i], b = this.Cells[i];
                if (a.GeometryDataFlags != b.GeometryDataFlags || !Same(a.WidthWithGap, b.WidthWithGap)
                    || !Same(a.HeightWithGap, b.HeightWithGap) || !ReferenceEquals(a.GeometryReference, b.GeometryReference)
                    || a.Geometry.Count != b.Geometry.Count) return false;
                for (int j = 0; j < a.Geometry.Count; j++)
                {
                    DxfStoredTableCellGeometry x = a.Geometry[j], y = b.Geometry[j];
                    if (!Same(x.TopLeftDistance, y.TopLeftDistance) || !Same(x.CenterDistance, y.CenterDistance)
                        || !Same(x.ContentWidth, y.ContentWidth) || !Same(x.ContentHeight, y.ContentHeight)
                        || !Same(x.Width, y.Width) || !Same(x.Height, y.Height) || x.StoredValue95 != y.StoredValue95) return false;
                }
            }
            return true;
        }

        private static bool Same(double first, double second) { return BitConverter.DoubleToInt64Bits(first) == BitConverter.DoubleToInt64Bits(second); }
        private static bool Same(Vector3 first, Vector3 second) { return Same(first.X, second.X) && Same(first.Y, second.Y) && Same(first.Z, second.Z); }
        private static void AddPoint(ICollection<DxfTag> tags, short code, Vector3 point)
        { tags.Add(new DxfTag(code, point.X)); tags.Add(new DxfTag((short)(code + 10), point.Y)); tags.Add(new DxfTag((short)(code + 20), point.Z)); }
        internal static void CheckFinite(double value, string parameter)
        { if (double.IsNaN(value) || double.IsInfinity(value)) throw new ArgumentOutOfRangeException(parameter, "Stored geometry must be finite."); }
        internal static void CheckFinite(Vector3 value, string parameter)
        { CheckFinite(value.X, parameter); CheckFinite(value.Y, parameter); CheckFinite(value.Z, parameter); }
    }
}

File: javascript/netDxf/Objects/DxfStoredTableGeometry.Edit.js
// Copyright (c) netDxf contributors. MIT License; see package LICENSE.
import { TableGeometryState } from '../../runtime/TableGeometryState.js';
import { TableSnapshot, TableHandleMap } from '../../runtime/TablePayload.js';
import { ConsumeManagedEnumerable } from '../../runtime/ManagedEnumerable.js';
import { ReferenceList } from '../../runtime/ReferenceList.js';
import { OrdinalIgnoreCaseKey } from '../../runtime/Collections.js';
import { DxfTag } from '../IO/DxfTag.js';
import { XDataCode } from '../XDataCode.js';
import { ArgumentException, ArgumentNullException, ArgumentOutOfRangeException, InvalidOperationException } from '../../runtime/Errors.js';
export function GeometryRecordTagCount(owner, payloadCount) {
  let count = payloadCount + 2;
  if (owner.ExtensionDictionary !== null) count += 3;
  const reactors = new Set(Array.from(owner.PersistentReactors).filter(item => item !== null).map(item => item.Handle === null ? null : OrdinalIgnoreCaseKey(item.Handle)));
  if (reactors.size) count += reactors.size + 2;
  for (const data of owner.XData.Values) { count++; for (const record of data.XDataRecord) count += record.Code === XDataCode.BinaryData ? Math.max(1, Math.trunc((record.Value.length + 126) / 127)) : 1; }
  return count;
}
function referenceText(cell) {
  const current = cell.GeometryReference === null ? '0' : cell.GeometryReference.Handle;
  return cell.ReferenceHandle !== null && BigInt('0x' + cell.ReferenceHandle) === BigInt('0x' + current) ? cell.ReferenceHandle : current;
}
const samePoint = (a, b) => Object.is(a.X, b.X) && Object.is(a.Y, b.Y) && Object.is(a.Z, b.Z);
function sameGeometry(owner, rows, columns, cells) {
  if (rows !== owner.RowCount || columns !== owner.ColumnCount || cells.length !== owner.Cells.Count) return false;
  return cells.every((a, i) => {
    const b = owner.Cells.get_Item(i);
    return a.GeometryDataFlags === b.GeometryDataFlags && Object.is(a.WidthWithGap, b.WidthWithGap) && Object.is(a.HeightWithGap, b.HeightWithGap) && a.GeometryReference === b.GeometryReference && a.Geometry.Count === b.Geometry.Count &&
      Array.from(a.Geometry).every((x, j) => { const y = b.Geometry.get_Item(j); return samePoint(x.TopLeftDistance, y.TopLeftDistance) && samePoint(x.CenterDistance, y.CenterDistance) &&
        ['ContentWidth', 'ContentHeight', 'Width', 'Height', 'StoredValue95'].every(key => Object.is(x[key], y[key])); });
  });
}
export function InstallTableGeometryEditing(Type) {
  Type.prototype.ReplaceGeometry = function(rowCount, columnCount, cells) {
    const state = TableGeometryState(this), fail = text => { throw new InvalidOperationException(text); };
    if (state.editing) { state.reentered = true; fail('TABLEGEOMETRY replacement cannot be reentered.'); }
    if (cells == null) throw new ArgumentNullException('cells');
    if (rowCount < 0 || rowCount > Type.MaximumPayloadTags) throw new ArgumentOutOfRangeException('rowCount');
    if (columnCount < 0 || columnCount > Type.MaximumPayloadTags) throw new ArgumentOutOfRangeException('columnCount');
    state.editing = true; state.reentered = false;
    try {
      let tagCount = 4; const replacement = [];
      ConsumeManagedEnumerable(cells, cell => {
        if (cell == null) throw new ArgumentException('A stored geometry cell cannot be null.', 'cells');
        tagCount += 5 + 11 * cell.Geometry.Count;
        if (tagCount > Type.MaximumPayloadTags) throw new ArgumentException('The requested geometry exceeds the record tag limit.', 'cells');
        replacement.push(cell);
      });
      if (state.reentered) fail('TABLEGEOMETRY replacement was reentered during enumeration.');
      if (!state.resolved || this.IsErased || this.Database === null || this.Database.Document !== state.source || state.source.GetObjectByHandle(this.Handle) !== this) fail('TABLEGEOMETRY must remain registered in its source document.');
      const errors = this.Database.Validate();
      if (errors.Count) fail('Cannot replace TABLEGEOMETRY in an invalid source database: ' + Array.from(errors).join('; '));
      if (GeometryRecordTagCount(this, tagCount) > Type.MaximumPayloadTags) throw new ArgumentException('The requested geometry and common metadata exceed the record tag limit.', 'cells');
      const references = new ReferenceList(), handles = new TableHandleMap();
      for (const cell of replacement) {
        const target = cell.GeometryReference; if (target === null) continue;
        if (target.Handle === null || state.source.StoredTableHandleTarget(target.Handle) !== target) throw new ArgumentException('Every geometry reference must be an actual registered identity in the source document.', 'cells');
        references.Add(target); handles.set(referenceText(cell), target);
      }
      if (sameGeometry(this, rowCount, columnCount, replacement)) return;
      const tags = [], add = (code, value) => tags.push(new DxfTag(code, value));
      const point = (code, value) => { add(code, value.X); add(code + 10, value.Y); add(code + 20, value.Z); };
      add(100, 'AcDbTableGeometry'); add(90, rowCount); add(91, columnCount); add(92, replacement.length);
      for (const cell of replacement) {
        add(93, cell.GeometryDataFlags); add(40, cell.WidthWithGap); add(41, cell.HeightWithGap); add(330, referenceText(cell)); add(94, cell.Geometry.Count);
        for (const value of cell.Geometry) { point(10, value.TopLeftDistance); point(11, value.CenterDistance); add(43, value.ContentWidth); add(44, value.ContentHeight); add(45, value.Width); add(46, value.Height); add(95, value.StoredValue95); }
      }
      const payload = TableSnapshot(tags), snapshot = TableSnapshot(replacement);
      state.payload = payload; state.cells = snapshot; state.references = references; state.handles = handles; state.rows = rowCount; state.columns = columnCount;
    } finally { state.editing = false; state.reentered = false; }
  };
}
